datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1446,18 +1446,54 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
 
             // Rules for Like
             Expr::Like(Like {
-                expr,
-                pattern,
+                expr: ref like_expr,
+                ref pattern, // Changed to ref pattern to avoid moving
                 negated,
-                escape_char: _,
-                case_insensitive: _,
-            }) if !is_null(&expr)
-                && matches!(
-                    pattern.as_ref(),
-                    Expr::Literal(ScalarValue::Utf8(Some(pattern_str))) if pattern_str == "%"
-                ) =>
-            {
-                Transformed::yes(lit(!negated))
+                escape_char,
+                case_insensitive,
+            }) => {
+                if let Expr::Literal(ScalarValue::Utf8(Some(pattern))) = pattern.as_ref()
+                {
+                    // Special case: pattern is just "%"
+                    if pattern == "%" && !is_null(like_expr) {
+                        return Ok(Transformed::yes(lit(!negated)));
```

Comment:
This is cool idea, but `is_null(like_expr)` checks makes it incorrect.
This function returns "is this always null". Applied to a column reference, it will return false, so this optimization will kick in, but the column values are still _nullable_.

The "more correct" way would be to

- optimize `WHERE col LIKE '%'` to `WHERE col IS NOT NULL`
- optimize `WHERE col NOT LIKE '%'` to `WHERE false`.

However, this simplification cannot be done **here**. It can only be done for top-level filter conjuncts, or when expr being simplified is immediately used as a condition (eg `IF(expr, ...)`). In other cases, this need to be NULL-preserving boolean logic:

```
trino> SELECT a, a LIKE '%', a NOT LIKE '%'
    -> FROM (VALUES NULL, '', 'x') t(a);
  a   | _col1 | _col2
------+-------+-------
 NULL | NULL  | NULL
      | true  | false
 x    | true  | false
(3 rows)
```

Reply:
The easiest solution would be to just drop this special case.
I.e. `expr LIKE '%'` would get simplified to `starts_with(expr, '')`

Reply:
See also https://github.com/apache/datafusion/issues/12637#issuecomment-2449352017